### DISCUSSION TITLE
Registration and Authentication using login and passwords

### DISCUSSION THREAD

**Discussion by biboobserota** [fief-dev/fief#209] [Q&A] [answered]
Hello, 
How to use your service for registration and authentication using login and passwords in FastAPI?
For example:
1. Using the endpoint `@app.post("/signup/")` I want to create 1 user using the following data: `username`, `email`, `password` and store this data in your cloud
2. Then using the endpoint `@app.post("/login/")` I want to login using for example `email`, `password`
3. How can I protect my endpoints using this authorization method and your library?
I did not find examples of how to implement this using Fief

**Comment by fief-bailiff** [comment]
Hail, @biboobserota 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to star our repository ⭐️

![star-fief](https://user-images.githubusercontent.com/1144727/192523552-dc2c31fa-ae94-42fa-9ab4-7d0ba51bb7b0.png)

Farewell!

**Reply by biboobserota** [reply]
done!

**Comment by frankie567** [comment, marked as answer]
Hi @biboobserota 👋 

Fief is not intended to be used directly as an authentication API. You need to think of it as a separate service which will handle all the login process. Your application will only receive a token you can use to identify the current user. We have a complete example for FastAPI here: https://docs.fief.dev/integrate/python/fastapi/

If you need more control and want to implement the authentication endpoints directly on your backend, then Fief might not be for you. You can have a look at dedicated libraries for this, like [FastAPI Users](https://fastapi-users.github.io).

**Reply by alexiamcdonald** [reply]
Heya @frankie567.

I'm also reading this post and I don't understand your vision since it's not clear on your website because I and others took it as a authentication API. 

`Fief is an open-source platform to manage users and authentication in your applications.` - [First line from your website homepage](https://docs.fief.dev/)

I am left with two questions now:
How are you meant to authenticate users if I as a backend system can't verify their passwords?
Why would you hand out an access token to a user who hasn't had their password verified?

Who/what is Fief intended for?

**Reply by frankie567** [reply]
Hey @alexiamcdonald 👋

What I meant in my previous message, is that Fief is designed to take care of all those aspects *for you*.

It means that, contrary to authentication libraries, it's not your backend that'll check for the password, the existence of the user and all the security aspects you could imagine.

You can think of Fief as a "microservice" which sole purpose is to authenticate your users. When users get back to *your* app, you have a valid access token, and can trust that they are correctly authenticated.

It's the same approach as very popular services like Auth0 or AWS Cognito.

Hope it clarifies things 🙂

**Reply by alexiamcdonald** [reply]
Hey @frankie567 

Thank you for your response.

Is it part of your vision to open up the API end points (specifically the signin endpoint) so Apps using Fief can use their own signin forms instead?

Or is there a plan to at least make the "Themes" be fully customisable so Fief's signin form fits with the connected app. I'm talking about roundness of input fields and shadow drop for example and icons, logos etc.

For instance, registering a user can be done already with the Fief API with our App's own form so I'm struggling to see why it wouldn't be the same (extended) for the signin as well.

**Reply by frankie567** [reply]
> Is it part of your vision to open up the API end points (specifically the signin endpoint) so Apps using Fief can use their own signin forms instead?

In the long term, yes, I would like to have that.

> Or is there a plan to at least make the "Themes" be fully customisable so Fief's signin form fits with the connected app. I'm talking about roundness of input fields and shadow drop for example and icons, logos etc.

Yes, I'll try to improve this over time. PR welcome on this matter :) 

> For instance, registering a user can be done already with the Fief API with our App's own form so I'm struggling to see why it wouldn't be the same (extended) for the signin as well.

I suppose you're doing this with the `POST /users/` endpoint of the Admin API, right? It works indeed if you rely on password authentication, but it's not possible to use it for social OAuth or future authentication methods we might introduce in the future (like magic links, passwordless, etc.).

**Reply by martincpt** [reply]
I would also like API endpoints for login and registration. Also, the fief client to have dedicated methods to perform these actions. It's pity it doesn't have these by default.

I understand FastAPIUsers is a dedicated library for that, but that's not a standalone microservice.

I believe these are crucial endpoints that shouldn't be omitted.

To be honest, I like Fief much better than Keycloak, but that's maybe a reason to stick with it for a while.
